File: software/src/sign_printer.py
from font5x7 import glyph
from sign import COLS

class SignPrinter:
    """Writes basic static text messages to the sign memory"""
    def __init__(self, sign):
        self.sign = sign
        self.screen_buff = bytearray(COLS)
        self.tmp_screen_buff = bytearray(COLS)

    def clear(self):
        self.sign.clear()
        self.sign.buffer_flip()
        return self

    def left(self, msg):
        buff = self.screen_buff
        msglen = SignPrinter.write_byte_array(msg, buff)
        index = max(0,int((COLS - msglen) / 2))
        self.sign.blit(2, buff, msglen) # first 2 columns are unaddressable
        self.sign.buffer_flip()

    def center(self, msg):
        buff = self.screen_buff
        msglen = SignPrinter.write_byte_array(msg, buff)
        index = max(0,int((COLS - msglen) / 2))
        self.sign.blit(index, buff, msglen)
        self.sign.buffer_flip()
        
    @micropython.native
    def center2(self, msg):
        buff = self.screen_buff
        msglen = SignPrinter.write_byte_array2(msg, buff)
        # index = max(0,int((COLS - msglen) / 2))
        # self.sign.blit(index, buff, msglen)
        self.sign.buffer_flip()

    def right(self, msg):
        buff = self.screen_buff
        msglen = SignPrinter.write_byte_array(msg, buff)
        index = max(0,int((COLS - msglen) / 2))
        self.sign.blit(COLS-msglen, buff, msglen) # first 2 columns are unaddressable
        self.sign.buffer_flip()

    def char_at_pos(self, ch, pos):
        glyph_cols = glyph(ch)
        self.sign.blit(pos, glyph_cols, len(glyph_cols))

    def _print(self, buff, index, len):
        self.sign.blit(index, buff, len)

    # writes the given message to the buff.
    # assumes that the buff is large enough to hold it
    def write_byte_array(msg, buff):
        offset = 0
        bufflen = len(buff)
        for i,ch in enumerate(msg):
            glyph_cols = glyph(ch)
            glyph_len = len(glyph_cols)
            buff[offset:offset+glyph_len] = glyph_cols
            offset = offset + glyph_len
            if(ch != ' ' and i < len(msg)-1 and offset < bufflen):
                buff[offset] = 0x00
                offset = offset + 1
            if(ch == ' '):
                offset = offset - 1
            if(offset >= bufflen):
                return bufflen
        return offset
    def write_byte_array2(msg, buff):
        msglen = len(msg)
        for i in range(0, msglen):
            # ch = 'a'
            # pass
            ch = msg[i]
            # glyph_cols = glyph(msg[i])
        # for i,ch in enumerate(msg):
        #     pass
        return 0

    # Fills a buffer that's full sign width with the message
    # Returns the actual inner content length, not the buffer length
    # TODO: Use alignment
    def fill_signbuff(self, msg, buff, align = 'center'):
        for i in range(0, len(buff)):
            buff[i] = 0x00
        tmpbuff = self.tmp_screen_buff
        msglen = SignPrinter.write_byte_array(msg, tmpbuff)
        # TODO: Use the alignment
        offset = int((COLS - msglen)/2)
        buff[offset:offset+msglen] = tmpbuff[0:msglen]
        return msglen

File: software/firmware/main/SignPrinter.cpp
#include <Arduino.h>
#include "sign_memory.h"
#include "font5x7.h"
#include "Sign.h"
#include "SignPrinter.h"

void SignPrinter::clear(){
  sign.clear();
}

// Prints a message to the sign at a given position
void SignPrinter::print(const char *msg, int pos, bool clear_first){
  if(clear_first) clear();

  uint8_t offset = pos;
  for(uint8_t i = 0; i < strlen(msg); i++){
    char ch = msg[i];
    GLYPH g = glyph(ch);
    sign.blit(offset, g.cols, g.length);
    offset += g.length;
    if(ch != ' '){
      sign.col(offset, 0x00);
      offset += 1;
    }
  }
}

// "prints" a message into a buffer. returns the actual length, in columns,
// used by the print.
uint8_t SignPrinter::print_mem(const char *str, uint8_t *buff, uint8_t bufflen, bool center){
  uint8_t index = 0;
  uint16_t str_cols_len = text_length(str);

  if(str_cols_len > bufflen) str_cols_len = bufflen;
  if(center){
    index = (bufflen - str_cols_len)/2;
  }
  uint8_t msglen = strlen(str);
  for(uint8_t i = 0; i < msglen; i++){
    if(index >= bufflen){
      return index;
    }
    GLYPH g = glyph(str[i]);
    for(uint8_t colnum = 0; colnum < g.length; colnum++){
      if(index >= bufflen) return index;
      buff[index++] = g.cols[colnum];
    }
    if((str[i] != ' ') && (i < msglen-1)){
      buff[index++] = 0x00;
    }
  }
  return index;
}

void SignPrinter::left(const char *msg, bool clear_first){
  print(msg, 2, clear_first); // first 2 cols aren't addressable
}

void SignPrinter::right(const char *msg, bool clear_first){
  uint16_t len = text_length(msg);
  uint8_t pos = (len >= SIGN_COLS) ? 0 : (SIGN_COLS-len);
  print(msg, pos, clear_first);
}

void SignPrinter::center(const char *msg, bool clear_first){
  uint16_t len = text_length(msg);
  uint8_t pos = (len >= SIGN_COLS) ? 0 : ((SIGN_COLS-len)/2);
  print(msg, pos, clear_first);
}
